kolibri/utils/conf.py

```diff
@@ -80,92 +78,126 @@
-#: first-run.
-config["AUTO_SEARCH_PLUGINS"] = []
-
-#: If a config file does not exist, we assume it's the first run
-config["FIRST_RUN"] = True
-
 conf_file = os.path.join(KOLIBRI_HOME, "kolibri_settings.json")
 
 
-def update(new_values):
-    """
-    Updates current configuration with ``new_values``. Does not save to file.
-    """
-    config.update(new_values)
-
-
-def save(first_run=False):
-    """Saves the current state of the configuration"""
-    config["FIRST_RUN"] = first_run
-    # use default OS encoding
-    with open(conf_file, "w") as kolibri_conf_file:
-        json.dump(config, kolibri_conf_file, indent=2, sort_keys=True)
-
-
-if not os.path.isfile(conf_file):
-    logger.info("Initialize kolibri_settings.json..")
-    save(True)
-else:
-    # Open up the config file and overwrite defaults
-    # use default OS encoding
-    with open(conf_file, "r") as kolibri_conf_file:
-        config.update(json.load(kolibri_conf_file))
-
-
-def autoremove_unavailable_plugins():
-    """
-    Sanitize INSTALLED_APPS - something that should be done separately for all
-    build in plugins, but we should not auto-remove plugins that are actually
-    configured by the user or some other kind of hard dependency that should
-    make execution stop if not loadable.
-    """
-    global config
-    changed = False
-    # Iterate over a copy of the list so that it is not modified during the loop
-    for module_path in config["INSTALLED_APPS"][:]:
-        if not module_exists(module_path):
-            config["INSTALLED_APPS"].remove(module_path)
-            logger.error(
-                (
-                    "Plugin {mod} not found and disabled. To re-enable it, run:\n"
-                    "   $ kolibri plugin {mod} enable"
-                ).format(mod=module_path)
-            )
-            changed = True
-    if changed:
-        save()
-
-
-def enable_default_plugins():
-    """
-    Enable new plugins that have been added between versions
-    This will have the undesired side effect of reactivating
-    default plugins that have been explicitly disabled by a user.
-    However, until we add disabled plugins to a blacklist, this is
-    unavoidable.
-    """
-    global config
-    changed = False
-    for module_path in DEFAULT_PLUGINS:
-        if module_path not in config["INSTALLED_APPS"]:
-            config["INSTALLED_APPS"].append(module_path)
-            logger.warning(
-                (
-                    "Default plugin {mod} not found in configuration. To re-disable it, run:\n"
-                    "   $ kolibri plugin {mod} disable"
-                ).format(mod=module_path)
-            )
-            changed = True
-
-    if changed:
-        save()
+# These values are encoded on the config dict as sets
+# so they need to be treated specially for serialization
+# and deserialization to/from JSON
+SET_KEYS = ["INSTALLED_APPS", "DISABLED_APPS"]
+
+
+class ConfigDict(dict):
```

Comment:
Non-blocking: normally my instinct is to [avoid subclassing](https://en.wikipedia.org/wiki/Composition_over_inheritance), particularly of built-ins types. In this case, I would personally have kept `config` a simple `dict`, and implemented the methods in this class (`update`, `save`, etc) as simple module-level functions which operate on `config`.